### ISSUE TITLE
SpringCGLIB proxy is not identified by ClassUtil

### ISSUE THREAD

**Issue by JayceJiaa** [spring-projects/spring-boot#12438]
Today I found it throws a ```Exception``` when start up in my demo project based on spring boot 2.0. It is a ```@Service``` annotated bean  used in my code by ```applicationContext.getBean(Iface)```(Iface is the interface my bean implemented). 

I followed into the source code in debug mode, and finally found that it returns false when invoke the method ```isAssignable``` in class ```org.springframework.util.ClassUtils```.

Comparing with my previous project based on spring 4.0, I found this method has call a native method ```java.lang.Class#isAssignableFrom```, which cannot identify the SpringCGLIB(in spring boot 2.0) proxy class but works well  with ```com.sun.proxy```

I do have no solution of this, is there a way? tks

**Comment by JayceJiaa** [comment]
The Exception is ```NoSuchBeanDefinitionException``` which I forgot to mention

**Comment by wilkinsona** [comment]
It sounds like you're being affected by Boot 2.0's change to using CGLib-based proxies rather than JDK proxies by default, but I can't tell why from what you've described thus far. Can you please provide a [minimal, complete, and verifiable example](https://stackoverflow.com/mcve) of your problem?

**Comment by JayceJiaa** [comment]
Please forgive me that I'm not allowed to submit code at work to github, which would take me more time to create a blank project and try to reproduce this error. If I make any progress, I will come back.

**Comment by wilkinsona** [comment]
Please do. If you'd like us to spend some time trying to help you, you're going to have to spend some time providing the information that we need in order to do so.

**Comment by JayceJiaa** [comment]
I'm sorry, I can't reproduce it in another project and everything just works well. 
In my debug process, I found that my problem is caused by the method ```isAssignable()``` in ```org.springframework.util.ClassUtils```.

![image](https://user-images.githubusercontent.com/22383871/37291843-83f7ac00-264a-11e8-921f-f4381066fe82.png)
When it goes into line 919 as shown in image, the value of expression ```lhsType.isAssignableFrom(rhsType)``` is false, which is a native method of jdk.

The debuger shows like this.
![image](https://user-images.githubusercontent.com/22383871/37292019-f24e9984-264a-11e8-8c5a-4a4c2d3e3f69.png)
 From the picture we can tell the native method cannot identify this implement relation.

And in my project to reduce the error, when it goes into this line, it works strangely good.
![image](https://user-images.githubusercontent.com/22383871/37292281-910ae096-264b-11e8-9fbd-7c51184d2f9b.png)

My jdk version is :
```
java version "1.8.0_51"
Java(TM) SE Runtime Environment (build 1.8.0_51-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)
```

I'm not sure I provided enough information but I really have no idea how I can do that

**Comment by wilkinsona** [comment]
Unfortunately, that doesn't help very much.

I can see in both cases that the proxies are CGLib proxies so my [guess above](https://github.com/spring-projects/spring-boot/issues/12438#issuecomment-372297912) was incorrect. I can also see that you're using DevTools and that, in both cases, the class loader of the proxy class (`rhsType`) is `RestartClassLoader@2950`. Perhaps the interface has been loaded by a different class loader in each case?

Please spend some more time trying to reproduce the problem in a small sample. Without an accurate description of the problem, or a sample that reproduces it, we're not going to be able to make much progress. We can't afford to spend time trying to diagnose something based on guesswork.

**Comment by JayceJiaa** [comment]
OK, I will try more. Thank you very much for you time and help. If I get that reproduced, I will inform you that.

**Comment by JayceJiaa** [comment]
I tried for a long time to reproduce this error and I failed, sorry for that.
But fortunately, your guess is correct.
It is caused by DevTools, maybe conflicting to the applicationContext in our own framework.
When I remove DevTools dependency, everything goes right. I'm not familiar with how the DevTools work, so I guess it's caused by the different ClassLoader used in DevTools.

And thank you very much for helping in the end!
